risc0/zkvm/src/prove/external.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2023 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    fs::File,
+    io::Write,
+    path::{Path, PathBuf},
+    process::{Command, Stdio},
+};
+
+use anyhow::{anyhow, bail, Result};
+use risc0_binfmt::MemoryImage;
+use tempfile::tempdir;
+
+use super::Prover;
+use crate::{ExecutorEnv, Receipt, Segment, SegmentReceipt, Session, VerifierContext};
+
+/// An implementation of a [Prover] that runs proof workloads via an external
+/// `r0vm` process.
+pub struct ExternalProver {
+    name: String,
+    r0vm_path: PathBuf,
+}
+
+impl ExternalProver {
+    /// Construct an [ExternalProver].
+    pub fn new<P: AsRef<Path>>(name: &str, r0vm_path: P) -> Self {
+        Self {
+            name: name.to_string(),
+            r0vm_path: r0vm_path.as_ref().to_path_buf(),
+        }
+    }
+}
+
+impl Prover for ExternalProver {
+    fn prove(
+        &self,
+        env: ExecutorEnv<'_>,
+        _ctx: &VerifierContext,
+        image: MemoryImage,
+    ) -> Result<Receipt> {
+        // TODO: deal with VerifierContext
+
+        log::debug!("Launching {}", &self.r0vm_path.to_string_lossy());
+        let temp_dir = tempdir()?;
+
+        let image_path = temp_dir.path().join("image");
+        let receipt_path = temp_dir.path().join("receipt");
+        {
+            let image_file = File::create(&image_path)?;
+            bincode::serialize_into(image_file, &image)?;
+        }
+
+        let mut cmd = Command::new(&self.r0vm_path);
+        cmd.arg("--image")
+            .arg(&image_path)
+            .arg("--receipt")
+            .arg(&receipt_path)
+            .stdin(Stdio::piped());
+
+        let mut child = cmd.spawn()?;
+
+        let mut stdin = child.stdin.take().ok_or(anyhow!("Failed to open stdin"))?;
+        std::thread::spawn(move || stdin.write_all(&env.input));
+
+        let status = child.wait()?;
+        if !status.success() {
+            bail!("Failed to run r0vm");
+        }
+
+        let receipt_bytes = std::fs::read(receipt_path)?;
```

Comment:
I think it would be nice to verify the receipt here using the verifier context

Reply:
Done